Wire message dispatch cleanup

da-kami opened the pull request:
Stability here we come ... 😬

da-kami committed:
Do not catch all when dispatching messages

This makes the code cleaner and better understandable.
When dispatching to `maker_cfd` actor using `taker_msg_channel` we use `send_async_safe` instead of relying on `send`'s ability to dispatch async without awaiting.

da-kami committed:
Dispatch all P2P connection related messages async to actors

Timeouts on message dispatch can cause the connection actors to be blocked, which can result in unwanted behaviour such as heartbeats not getting sent by the maker.
Receiving messages on the P2P connection should never lead to blocking behaviour, thus we dispatch all messages async to the actors.
This removes the previously introduced timeouts.
Additionally, this harmonizes message handling in the taker and maker connection which should make it easier to reason about the code.

As a result of this change, we don't have message handlers the return a result.
This is actually wanted, because the caller was not able to handle the result in a meaningful way.
Errors when processing a message are now handled when processing the message in the respective actor and result in error logs - which is basically the same behaviour as before, only that the logging was handled by the caller.

da-kami committed:
Remove unused `send_fallible` from `xtra` extension

With aligning the message dispatch to the actors we stopped using `send_fallible` and can remove this functionality.

klochowicz commented on daemon/src/maker_cfd.rs:
```diff
@@ -691,20 +691,12 @@ where
-            wire::TakerToMaker::HelloV2 { .. } => {
-                unreachable!("The HelloV2 message is not sent to the cfd actor")
-            }
-            wire::TakerToMaker::Unknown => {
-                // Ignore unknown message to be forwards-compatible.
```
why don't you want to be forwards-compatible??? you might have gone too far with combining them :)

da-kami replied:
I think this comment was misleading. 
I think the decision to process / not process a message should *not* be taken in `maker_cfd` but in the connection handling actor, i.e. `maker_inc_connections`. That's why I moved it there and made all message that we don't care about in `maker_cfd` `unreachable`. 
Am I missing something?

klochowicz replied:
fair enough! I missed this, it makes sense to squash it there, instead of moving it to this actor.

klochowicz commented on daemon/src/maker_cfd.rs:
```diff
@@ -691,20 +691,12 @@ where
                 }
             }
-            wire::TakerToMaker::RolloverProtocol { .. } => {
-                unreachable!("This kind of message should be sent to the rollover_maker::Actor`")
-            }
-            wire::TakerToMaker::Protocol { .. } => {
-                unreachable!("This kind of message should be sent to the `setup_maker::Actor`")
```
nit: the previous error messages were better (at least for me), but I don't mind that much

da-kami replied:
Same as my comment above - I don't think these different `unreachable` statements helped.
Eventually we could think about moving away from the `enum`. This code should just not be in `maker_cfd` at all - it is not concerned with these messages.

klochowicz replied:
as said above, don't mind too much with this one 🙃  it was just expressing an opinion about the contents of `unreachable` - it actually used to mention which actor handles it.

klochowicz commented on daemon/src/maker_inc_connections.rs:
```diff
@@ -435,7 +432,9 @@ impl Actor {
                             msg,
                         })
                         .await
-                        .context("`maker_cfd::Actor` is disconnected")?;
```
nice catch, fixing the misleading context! 👍

klochowicz commented on daemon/src/maker_inc_connections.rs:
```diff
@@ -494,40 +493,39 @@ impl Actor {
 
         use wire::TakerToMaker::*;
         match msg.msg {
-            Protocol { order_id, msg } => match self.setup_actors.get_connected(&order_id) {
```
did you mean to remove the `get_connected()` check from here? This doesn't seem like it should be part of the changeset (if anything, you can use `send_async()` inside, after finding the appropriate setup actor, instead of dispatching to all of them).

da-kami replied:
As far as I can tell this `get_connected` check will be implied when we use `send_async` on the `AddressMap` directly. This `get_connected` check here is legacy code for me. We might even be able to *not* expose `get_connected`. I can have a look at that.

klochowicz replied:
you are right! I didn't look much into `AddressMap` before, we're looking at the keys whenever we `send` as well, so that part was not needed there beforehand either....
We can remove that method altogether. :)

da-kami replied:
I added another commit that removes the `get_connected` check because it is unused now :)

klochowicz reviewed (changes requested):
concept ACK, a few things highlighted in the comments perhaps could be reverted - other than that, I'm keen to try it out!

luckysori commented on daemon/src/maker_cfd.rs:
```diff
@@ -691,20 +691,12 @@ where
-            wire::TakerToMaker::HelloV2 { .. } => {
-                unreachable!("The HelloV2 message is not sent to the cfd actor")
-            }
-            wire::TakerToMaker::Unknown => {
-                // Ignore unknown message to be forwards-compatible.
```
🔧 I would keep the behaviour of the `Unknown` variant as is. We don't want to panic if we receive an unknown message.

da-kami replied:
Same as in the reply to @klochowicz  - I think the `maker_cfd` actor should *not* be in charge of this decision. This decision should be taken in `maker_inc_connections`. I think this code was actually misleading and is the result of legacy code that was not changed in `maker_inc_connections`.

luckysori replied:
I'm content with not making this a special case as long as we only call `unreachable`  on `cfg!(debug_assertions)`.

da-kami replied:
OK with the fixup?

luckysori commented on daemon/src/maker_cfd.rs:
```diff
@@ -691,20 +691,12 @@ where
+            | wire::TakerToMaker::Hello(_)
+            | wire::TakerToMaker::HelloV2 { .. }
+            | wire::TakerToMaker::Unknown => {
+                unreachable!("Message {} is not dispatched to this actor", msg.name())
```
🔧 We consider this unreachable, but it's still risky to have `unreachable`s in production. What do you think about this?

```rust
if cfg!(debug_assertions) {
    unreachable!("Message {} is not dispatched to this actor", msg.name())
}
```

I reckon it's safer to just ignore unexpected messages in general. We won't need to deal with this situation once we migrate to the new network layer, but that's still weeks away.

klochowicz replied:
Really like the suggestion - perhaps we could go throughout our codebase and deal with it with one sweep? maybe a macro that at least logs a warning if this happens in production, and panics in debug?

luckysori replied:
Hehe it's inspired by your idea of using `debug_assert` 😛

I think there's only one other place where we use `unreachable` in production, but I like the idea of introducing a macro that panics in debug but logs in production.

da-kami replied:
I am happy to remove the `unreachable` here, I am not a fan of it! 
I like the proposal, in addition we could log a `warn` in production.

luckysori commented on daemon/src/maker_inc_connections.rs:
```diff
@@ -530,8 +530,21 @@ impl Actor {
+                let _ = self.taker_msg_channel.send_async_safe(msg).await;
+            }
+            Hello(_) | HelloV2 { .. } => {
+                // No dispatching because these messages are handled when upgrading the connection
```
🔧 We could follow the same pattern I suggested above here.

luckysori commented on daemon/src/connection.rs:
```diff
@@ -439,82 +436,46 @@ impl Actor {
-                        tracing::warn!(%order_id, "Contract setup not started: Contract setup took too long to process Accept")
+                    Ok(_) => {}
+                    Err(NotConnected(_)) => {
+                        tracing::error!(%order_id, "No active setup actor");
```
🔧 Can replace with `if let Err(e)`.

luckysori replied:
And below.

luckysori commented on daemon/src/setup_maker.rs:
```diff
@@ -219,14 +219,20 @@ impl Actor {
-        Ok(())
+        if let Err(e) = sender.send(msg).await {
+            tracing::error!("Failed to dispatch contract setup message to task: {e}")
+        }
```
🔧 To reduce the number of times you have to explicitly handle errors, you can extract the body of this method into a non-`xtra` method on `Actor` which returns a `Result`. Then you can handle that result here with a single `if let Err(e)`.

And then you can include the whole error stack with `{e:#}`.

luckysori commented on daemon/src/setup_maker.rs:
```diff
@@ -219,14 +219,20 @@ impl Actor {
+        let mut sender = match self.setup_msg_sender.clone() {
+            Some(sender) => sender,
+            None => {
+                tracing::error!(
```
🔧 I think we want most of the logs in this PR to remain as `warn`.

luckysori reviewed (approved):
LGTM except for some details.

klochowicz reviewed (approved):
nice one!!!

da-kami committed:
Removed unused `get_connected` check

With harmonizing the message handling in the P2P connection actors we do not need this message anymore, because we only use `send_async` which does this check implicitly by mapping errors to `NotConnected`.
Note that `send_async` ensures a message handler that does not return a `Result`, so the only error that can happen is `NotConnected`.

luckysori reviewed (approved):
Amazing! Thanks for taking care of the suggestions.

klochowicz reviewed (approved):
LGTM! Thank you for taking care of this!
